[blog] Date and time pickers revamped (#35486)
mui/material-ui: docs/pages/blog/nextgen-pickers.md

```jsx
import { Unstable_DateField as DateField } from '@mui/x-date-pickers/DateField';

<DateField label="My first field" />;
```

They are also included by default on the new pickers.

```jsx
import { Unstable_NextDatePicker as NextDatePicker } from '@mui/x-date-pickers/NextDatePicker';

<NextDatePicker label="My first v6 picker" />;
```

## Enhanced keyboard usability

With the new fields, not only can you type the date as a text, but you can use your arrow keys to quickly navigate through each piece of date and time values.
You can edit them atomically, without, for instance, leaking the year into the month, a common occurrence with masked inputs.

<video style="margin-bottom: 24px;" autoplay muted loop playsinline controls>
  <source src="/static/blog/nextgen-pickers/keyboard-nav.mp4" type="video/mp4" />
</video>

For a quick comparison, let's check out, for instance, how the two different approaches behave on the most typical scenarios.

<video style="margin-bottom: 24px;" autoplay muted loop playsinline controls>
  <source src="/static/blog/nextgen-pickers/fields-common-usecases.mp4" type="video/mp4" />
</video>

> Try out the live demo of the `DateField` below.

Author: Not directly related to the blog post, but I believe we should somehow improve the date field to move across the date pieces automatically when there's valid input. Something like [spectrum](https://react-spectrum.adobe.com/react-spectrum/DatePicker.html) does. I'm afraid this we'll be a reason people will stumble when using the component, especially for the first time.

> Try out the live demo of the `DateField` below.

Author: Not directly related to the blog post, but I believe we should somehow improve the date field to move across the date pieces automatically when there's valid input. 
Something like [spectrum](https://react-spectrum.adobe.com/react-spectrum/DatePicker.html) does. I'm afraid this we'll be a reason people will stumble when using the component, especially for the first time.

cc @flaviendelangle

Reviewer: I think we can add a follow up section for the fields, with our 2-3 main improvements ideas for the future.
Could be

- Auto section switching
- Week days support
- Smart pasting (like Retool)

That way we show that the components are not perfect, but that we are moving fast on them.

<iframe
  src="https://codesandbox.io/embed/date-field-demo-pb87v0?fontsize=12&hidenavigation=1&module=%2F"
  style="width:100%; height:200px; border:0; border-radius: 4px; overflow:hidden;"
  allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
  sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
></iframe>

## Other Significant improvements

Now that you've seen what's fundamentally different with the pickers, let us brief you on some other noteworthy refinements the components are getting.

> - **Customization of elements and sub-components in the Clock and Calendar**

Author: Ok, I wrote like that because that's what we would call the views, but I see how this can mix the concepts.

```suggestion
- **Customization of layout and internal components**
```

> - **Customization of elements and sub-components in the Clock and Calendar**

Reviewer: The new title is fine for me
Maybe we could link to the new layout doc when published

---

> title: A fresh approach for date and time inputs

Reviewer 2: I feel that it would be much better to tighten new feature announcements like this one to actual version releases. It makes it very clear which version includes this and gives a reason for people to share a specific release.

description: Check out what's new with the next-gen pickers in earlier v6 versions.
date: 2022-12-20T00:00:00.000Z
authors: ['josefreitas']
tags: ['MUI X', 'News']
---

There's a lot of exciting news in MUI X v6, but there's hardly anything comparable to the revamp we're delivering for the [Date and Time Pickers](https://next.mui.com/x/react-date-pickers/getting-started/).

> It's the result of our research and effort to improve these inputs' usability and overall customizability.

Reviewer 3: ```suggestion
This is the result of our ongoing efforts to improve the usability and customizability of these components.
```

We kept the best aspects of the previous versions, and we're fleshing out the package with several improvements and one fundamental change: a new input type to replace the masked text fields.

> ## The new date and time fields

Reviewer 2: As a general rule, I think that the blog posts that get people excited always focus the majority of the attention on the problem, they never put the solution at the center.

In this post, I think that we could focus more on seeing a clear demonstration of the problems. For example, this is a horrible UX, we could cover it:

https://user-images.githubusercontent.com/3165635/210186164-ef38aecf-d0e4-412f-914f-f71b9e7e3ec8.mov

https://mui.com/x/react-date-pickers/date-picker/#basic-usage WUT is going on with the year? WUT is the field in the error stage, I'm still editing?

> This new input type we're talking about is a set of components shortly named [Fields](https://next.mui.com/x/react-date-pickers/fields/).
> They are the base for the new pickers and a revolution in terms of keyboard usability.

Reviewer 3: ```suggestion
These new input components are called [Fields](https://next.mui.com/x/react-date-pickers/fields/), and they're revolutionary when it comes to navigating through the Picker components with a keyboard.
```

Reviewer 3: Yeah, that's a good point. "Revolutionary" feels like too much of an exaggeration. Maybe something like this instead?

```suggestion
These new input components are called [Fields](https://next.mui.com/x/react-date-pickers/fields/), and they offer massive improvements for navigating through the Picker components with a keyboard.
```

Reviewer: In term of keyboard navigation, I think you could say that it's x10 better, given that the mask input had no real keyboard navigation.

But for the component as a whole, I do agree that "revolution" might put the bar a little to high.

Reviewer 3: Ah ok, I misinterpreted what's so great about them. 😁

How about this?

```suggestion
These new input components are called [Fields](https://next.mui.com/x/react-date-pickers/fields/), and they provide massive improvements to the user experience of editing the Pickers with a keyboard.
```

Reviewer: That's fine for me.

Strictly speaking the Field being usable on their own, the following is maybe more correct:

```suggestion
These new input components are called [Fields](https://next.mui.com/x/react-date-pickers/fields/), and they provide massive improvements to the user experience of editing the value with a keyboard.
```

Reviewer 2: > In term of keyboard editing, I think you could say that it's x10 better, given that the mask input had no keyboard editing apart from letters and digits.

@flaviendelangle Ok fair, but I think that the sentiments stay, to me, revolutionary would involve turning things around. But I feel that we are taking steps toward an existing non-controversial UX pattern. The latest proposal sounds great.

> In contrast with the masked inputs that only display simple text behind a formatting mask, and were often cumbersome, especially when editing dates, the new fields are smart and specialized in editing date and time numbers.

Reviewer 3: This long run-on sentence is best split up in two.

```suggestion
The earlier masked inputs were often quite cumbersome to use—especially when editing dates.
The new Fields, by contrast, are optimized for the best possible editing experience.
```

Reviewer: Shouldn't we use `Masked inputs` + `Fields` or `masked inputs` + `fields` ?

Reviewer 3: @flaviendelangle which of these does "masked inputs" refer to? It's not clear to me in context, but this would be the difference in terms of style:
a. MUI component = Masked Inputs
b. general concept = masked inputs

Reviewer: If that's not clear for you, then we probably need to rephrase.

The input of the old pickers was based on a mask library called [rifm](https://www.npmjs.com/package/rifm). It was not usable as a standalone component so its name was not really normalized (the private component is called `KeyboardDateInput`)

The input of the new pickers are based on the Field components, which are usable as standalone components.

But from a user perspective, `mask` is a term that people only encountered if they tried to customize the behaviors. It's not a word we are referring a lot to in the doc like `field` will be.

Maybe we could rephrase to something like: 

```suggestion
In contrast with the previous generation of pickers, whose input editing was based on a [mask approach](https://www.npmjs.com/package/rifm) and were often cumbersome, the new fields are smart and specialized in editing date and time values.
```

I am replacing `numbers` by `values` (if you have a better idea) because we support letter sections like full letter months (`January`) as well.

And to be honest I don't understand the `especially when editing dates`, what else are they doing than editing dates ? :angel:

Author: > And to be honest I don't understand the especially when editing dates, what else are they doing than editing dates ?

Just a quick note on this one: editing vs. inserting

Reviewer: Oh OK, in that case maybe "editing" is not the right word because we are also using it to describe the whole process (the initial filling and the modification of already existing values).

Maybe `modifying a value` would describe it better.

> It can recognize what's a month, day, or whatever piece of the date you are currently editing and act accordingly.

Reviewer 3: ```suggestion
They're "smart" in the sense that they can recognize when a user is inputting any given piece of time data—a month, a day, a year—and respond accordingly.
```

> For example, watch what happens when you edit the month with arrow keys.

Reviewer 3: ```suggestion
For example, watch what happens when you edit the month using the arrow keys:
```

> <video style="margin-bottom: 24px;" autoplay muted loop playsinline controls>

Reviewer 2: The size of the video is off. It feels bigger than in real life. It feels better with:

```suggestion
<video style="margin-bottom: 24px; max-width: 615px" autoplay muted loop playsinline controls>
```

x the other videos to keep them 1:1 in scale.

Reviewer 2: Since the video is relatively short, I thin that we could do
```suggestion
<video style="margin-bottom: 24px;" autoplay muted loop playsinline>
```

https://www.notion.so/mui-org/Blog-247ec2bff5fa46e799ef06a693c94917#cdbb73821e23409c94f184dc4f5d2e61

x the other relevant videos

Reviewer 2: Do you know why there is this border on the recordings?

<img width="729" alt="Screenshot 2023-01-01 at 23 34 20" src="https://user-images.githubusercontent.com/3165635/210186444-f6ebed3a-9904-47f0-b9ab-765f7806fbbe.png">

Author: @oliviertassinari,
I'm re-recording the videos for this blog post and, meanwhile, trying to understand when the border happens. 
Initially, I thought it could be an artifact of a wrong aspect ratio or height set with an odd number, but none of those were confirmed correct; 

I couldn't get a consistent output without the border from a custom-sized screen recording with QuickTime, so now I'm removing the borders with an extra FFmpeg step to cut one pixel from the top and one from the bottom.

`ffmpeg -i masked-input.mp4 -filter:v "crop=in_w:in_h-2" -c:a copy masked-input_crop.mp4`

  <source src="/static/blog/nextgen-pickers/intelligent-field.mp4" type="video/mp4" />
</video>

As you increase the month, the component automatically watches for valid values throughout each piece of your date input, like you're navigating in a calendar.

> The Fields are a perfect match to the calendar view, but they are available as stand-alone components, and you can import them from the latest v6 package.

Reviewer 3: ```suggestion
The Fields are a perfect match to the calendar view, but they're available as standalone components, and you can import them from the latest v6 package.
```

Reviewer: > to the calendar views

I'd keep the plural here

```jsx

> import { Unstable_DateField as DateField } from '@mui/x-date-pickers/DateField';

Reviewer: You can remove all the `Unstable` from the blog post for both the fields and the pickers.

For the field, they don't have the prefix anymore on `next` (will be released this week).

For the pickers, we are renaming it to their final name before releasing this blog post

```suggestion
import { DateField } from '@mui/x-date-pickers/DateField';
```

<DateField label="My first field" />;
```

> They are also included by default on the new pickers.

Reviewer 3: ```suggestion
They're also included by default with the new Pickers.
```

```jsx

> import { Unstable_NextDatePicker as NextDatePicker } from '@mui/x-date-pickers/NextDatePicker';

Reviewer: ```suggestion
import { DatePicker } from '@mui/x-date-pickers/DatePicker';
```

> <NextDatePicker label="My first v6 picker" />;

Reviewer: ```suggestion
<DatePicker label="My first v6 picker" />;
```

```

## Enhanced keyboard usability

> With the new fields, not only can you type the date as a text, but you can use your arrow keys to quickly navigate through each piece of date and time values.

Reviewer 3: ```suggestion
With the new Fields, not only can you type the date as text, but you can also use your arrow keys to quickly navigate through each date and time value.
```

> You can edit them atomically, without, for instance, leaking the year into the month, a common occurrence with masked inputs.

Reviewer 3: ```suggestion
You can edit them atomically—for instance, without leaking the year into the month, which is a common occurrence with masked inputs.
```

<video style="margin-bottom: 24px;" autoplay muted loop playsinline controls>
  <source src="/static/blog/nextgen-pickers/keyboard-nav.mp4" type="video/mp4" />
</video>

> For a quick comparison, let's check out, for instance, how the two different approaches behave on the most typical scenarios.

Reviewer 3: ```suggestion
For a quick comparison, let's check out how the two different approaches behave in some typical scenarios.
```

<video style="margin-bottom: 24px;" autoplay muted loop playsinline controls>
  <source src="/static/blog/nextgen-pickers/fields-common-usecases.mp4" type="video/mp4" />
</video>

> Try out the live demo of the `DateField` below.

Reviewer 3: ```suggestion
Try out the live demo of the Date Field component below:
```

<iframe
  src="https://codesandbox.io/embed/date-field-demo-pb87v0?fontsize=12&hidenavigation=1&module=%2F"
  style="width:100%; height:200px; border:0; border-radius: 4px; overflow:hidden;"
  allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
  sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"

> ></iframe>

Reviewer 2: If we want to keep this codesandbox, I think that we should freeze the versions. I would expect it to break before the end of the year:

<img width="454" alt="Screenshot 2023-01-01 at 23 35 41" src="https://user-images.githubusercontent.com/3165635/210186458-3fd8130a-0a46-433d-96ad-95b7d475f4ac.png">

We could also make the codesandbox part of the MUI org so anybody in the team can make edits without changing the sandbox. Steps:

1. open https://www.notion.so/mui-org/Accounts-945f0ccb9f7749ccb9e84b9eb9b13ec7#04a515471469433daa20fb1a44b1525f and follow the link to join the org.
2. fork 

<img width="291" alt="Screenshot 2023-01-01 at 23 38 07" src="https://user-images.githubusercontent.com/3165635/210186519-06597044-223a-4905-88b8-3a717fe9ca93.png">

Author: I froze the versions on the sandbox, but I can't fork it with MUI.
(The pickers still need to get updated when beta is actually out)

<img width="294" alt="image" src="https://user-images.githubusercontent.com/550141/213293886-cead5b9b-afbf-42cb-bc3d-9a7bb670bff4.png">

Reviewer 2: > I froze the versions on the sandbox

Great, it was the most important

> I can't fork it with MUI

We are supposed to be under an Open Source plan which gives us unlimited editors. They got something wrong, I'm letting them know.

### Fields roadmap

> The Fields are still in alpha, and there are many improvements on the way. The following list describes some of the most noticeable.

Reviewer 3: ```suggestion
The Fields are still in alpha, and there are many improvements on the way. The following list describes some of the most notable features planned.
```

> - **Auto section switching**

Reviewer 3: ```suggestion
- **Automatic section switching**
```

>   So users don't need to manually navigate to the next section of the date after fulfilling the current one.

Reviewer 3: ```suggestion
  Users won't need to manually navigate to the next section of the date after filling in the preceding field.
```

Reviewer: Same, can be removed

- **Smart pasting**

>   Date strings pasted on Fields are parsed and matched with the sections of the date.

Reviewer 3: Future tense since this is coming soon.

```suggestion
  Date strings pasted in Fields will be parsed and matched with the sections of the date.
```

- **Support days of the week**

>   Support selecting a date by navigating days of the week with arrow keys, or text editing.

Reviewer 3: ```suggestion
  Users will be able to select a date by navigating through the days of the week with arrow keys, or by editing text.
```

Reviewer: Maybe simply:

```suggestion
  Users will be able to select a day in the week just like the can select a month.
```

> We'd love to see more of your [feedback](https://github.com/mui/mui-x/issues/new/choose) to help us improve these components even further.

Reviewer 3: ```suggestion
We'd love to hear your [feedback](https://github.com/mui/mui-x/issues/new/choose) to help us improve these components even further.
```

> ## Other Significant improvements

Reviewer 3: ```suggestion
## Other significant improvements
```

> Now that you've seen what's fundamentally different with the pickers, let us brief you on some other noteworthy refinements the components are getting.

Reviewer 3: ```suggestion
Now that you've seen what's fundamentally changed with the Pickers, let's briefly review some other noteworthy improvements.
```

- **Customization of layout and internal components**

>   Combining the concept of slots and grid layout, you now can rearrange, extend, and customize most of the components used internally by the pickers views, and make them perfectly suited to your use cases.

Reviewer 3: ```suggestion
  Combining the concept of slots with the grid layout, you now can rearrange, extend, and customize most of the components used internally by the Pickers' views.
```

- **Default render input**

  You don't need to provide the usual `TextField` on `renderInput` anymore unless you need it customized.
  In which case you can do it by:

  1. Customizing through props via `field` and `input` slots ([codebox example](https://codesandbox.io/s/customizing-fields-with-props-o66r1c?file=/demo.tsx)).
  2. Build a new field component using the Field's headless API (documentation coming soon).

>   3. Use an entirely [custom text field](https://next.mui.com/x/react-date-pickers/date-picker/#custom-input-component) to suit your use cases.

Reviewer 3: ```suggestion
  3. Use an entirely [custom text field](https://next.mui.com/x/react-date-pickers/date-picker/#custom-input-component) to suit your use case.
```

- **Drag to edit**

>   Editing a date range is even easier now with the new interface using drag and drop. Change `start` and `end` dates at will.

Reviewer 3: ```suggestion
  Editing a date range is even easier now with the new drag-and-drop interface. Change `start` and `end` dates at will.
```

Reviewer: Is their a reason to say "`start`" instead of "start" ?
It's still a valid sentence without referring to the variable.

<video style="margin-bottom: 24px;" autoplay muted loop playsinline controls>
  <source src="/static/blog/nextgen-pickers/edit-drag.mp4" type="video/mp4" />

> </video>

Reviewer 2: I think that the video is missing https://www.notion.so/mui-org/Blog-247ec2bff5fa46e799ef06a693c94917#7265c594d41e4906a521eb1a05caef53. At first, I thought that there was a bug with the selection state UI, but it's only because I didn't see when the right click was pressed. It feels clearer like this:

https://user-images.githubusercontent.com/3165635/210185968-b527d02b-a7af-451f-9e9a-aa3f3f0d4bf7.mov

- **Range shortcuts**

>   Add quick and customizable shortcuts to your users. Choose to display them on left, right, bottom or top.

Reviewer 3: ```suggestion
  Add quick and customizable shortcuts for your users. Choose to display them on the left, right, bottom or top.
```

<img src="/static/blog/nextgen-pickers/date-range-shortcuts.png" style="width: 692px; margin-bottom: 24px;" loading="lazy" alt="Date Range shortcuts." />

## Should we retire the masked inputs?

We believe that after using the date fields, you'll have no reason to come back to the masked input approach.
It has a few irreparable usability issues, and it sets some hard limits on what we can do with the input.

> However, they are still available during the pre-release, and you can find their documentation under the documentation pages prefixed with 'Legacy'.

Reviewer 3: ```suggestion
However, they are still available during the pre-release, and in the documentation they now have the term "Legacy" appended to their names.
```

> The plan is to replace them entirely on the v6 stable release. But in the end, it's all up to you, and we'd love to listen to what you think.

Reviewer 3: I would be a little more direct about this.

```suggestion
Our plan is to replace them entirely in the stable release of v6, but if that will prove to be a serious problem for you, then you still have a chance to influence our roadmap.
```

> If you want to voice your opinions or get involved, please consider giving us a [user interview](https://forms.gle/vsBv6CLPz9h57xg8A).

Reviewer 3: ```suggestion
If you want to voice your opinions or get involved, please consider scheduling a [user interview](https://forms.gle/vsBv6CLPz9h57xg8A) with us.
```

> You can also, and always, get in touch by opening new bug reports or feature requests in our [GitHub repository](https://github.com/mui/mui-x/issues/new/choose).

Reviewer 3: ```suggestion
As always, you can get in touch by opening new bug reports or feature requests in our [GitHub repository](https://github.com/mui/mui-x/issues/new/choose).
```

Moreover, notice on the following example that as we increase the month, the component automatically watches for valid values throughout each piece of your date input, like you're navigating in a calendar.

<video style="margin-bottom:5px; max-width:608px" autoplay muted loop playsinline >
  <source src="/static/blog/nextgen-pickers/smart-field.mp4" type="video/mp4" />
</video>

For a quick comparison, let's check out how the old and new approaches behave in typical scenarios.

<video style="margin-bottom:24px" autoplay muted loop playsinline >
  <source src="/static/blog/nextgen-pickers/quick-comparison-fields.mp4" type="video/mp4" />
</video>

Please, try it out for yourself in the live demo below:

<iframe
  src="https://codesandbox.io/embed/date-field-demo-pb87v0?fontsize=12&hidenavigation=1&module=%2F"
  style="width:100%; height:200px; border:0; border-radius: 4px; overflow:hidden;"
  allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
  sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
></iframe>

### Fields roadmap

> The Fields are still in alpha, and there are many improvements on the way. The following list describes some of the most notable features planned.

Reviewer: ```suggestion
The Fields are still in beta, and there are many improvements on the way. The following list describes some of the most notable features planned.
```

- **Smart pasting**

  Date strings pasted in Fields will be parsed and matched with the sections of the date.

- **Support days of the week**

>   Users will be able to select a day in the week just like the can select a month.

Reviewer: Should we link to the github issues of those two ?

Author: I've added the links.
